File: Python_Solutions/Maximum_Depth_of_Binary_Tree.py
"""
Leetcode Challenge: Maximum Depth of Binary Tree

Given the root of a binary tree, return its maximum depth.
A binary tree's maximum depth is the number of nodes along the 
longest path from the root node down to the farthest leaf node.

6/12/24
Mark Robuck
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    """
    Find the maximum depth of a binary tree.
    Args:
        root(TreeNode): the root node of the tree
    Returns:
        int: the maximum depth of the tree
    """
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        left_depth = 1
        right_depth = 1
        #base case of depth 0
        if root is None:
            return 0
        #recursive case, find the depth of left and right subtrees
        left_depth += self.maxDepth(root.left)
        right_depth += self.maxDepth(root.right)
        #max depth is the greater of the two subtree depths
        return max(left_depth, right_depth)

File: Java_Solutions/Maximum_Depth_of_Binary_Tree.java
/*
* Leetcode Challenge: Maximum Depth of Binary Tree
*
* Given the root of a binary tree, return its maximum depth.
* A binary tree's maximum depth is the number of nodes along the 
* longest path from the root node down to the farthest leaf node.
*
* 6/12/24
* Mark Robuck
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /*
    * Find the maximum depth of a binary tree.
    * @params root, the root node of the tree
    * @return, the maximum depth of the tree as an int
    */
    public int maxDepth(TreeNode root) {
        int leftDepth = 1;
        int rightDepth = 1;
        //base case of depth 0
        if (root == null) {
            return 0;
        }
        //recursive case, find the depth of left and right subtrees
        leftDepth += this.maxDepth(root.left);
        rightDepth += this.maxDepth(root.right);
        //max depth is the greater of the two subtree depths
        return Math.max(leftDepth, rightDepth);       
    }
}
